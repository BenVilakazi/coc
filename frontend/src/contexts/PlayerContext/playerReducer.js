function joinLobby(state) {
  return {
    ...state,
    gameState: 'pending-connection',
    message: {
      big: 'Connecting to Lobby',
      small: 'Please wait',
    },
    loading: [...state.loading, 'joining-lobby'],
  };
}

function submitCards(state) {
  return {
    ...state,
    gameState: 'submitting-cards',
    message: {
      big: 'Submitting your cards',
      small: 'Please wait',
    },
    loading: [...state.loading, 'submitting-cards'],
  };
}

function update(state, payload) {
  const { removeLoading, ...newData } = payload;
  return {
    ...state,
    ...newData,
    loading: removeLoading
      ? state.loading.filter((loadingVal) => loadingVal !== removeLoading)
      : state.loading,
  };
}

function errorDisconnect(state) {
  return {
    ...state,
    gameState: 'error',
    message: {
      big: 'AN ERROR OCCURRED',
    },
  };
}

function receiveWhiteCards(state, payload) {
  const { cards, selectCardCount } = payload;

  return {
    ...state,
    gameState: 'player-select',
    cards,
    selectCardCount,
  };
}

function lobbyClosed(state) {
  return {
    ...state,
    gameState: 'lobby-closed',
    message: {
      big: 'THE LOBBY HAS BEEN CLOSED',
      small: "You don't have to go home, but you can't stay here",
    },
  };
}

function noLobby(state) {
  return {
    ...state,
    gameState: 'error',
    message: {
      big: 'Lobby does not exist',
      small: 'We understand, spelling is pretty hard.',
    },
    loading: [],
  };
}

function removeDisconnectedPlayersCard(state, { playerId }) {
  if (state.submittedCards && state.submittedCards.length > 0) {
    const cardRemoved = state.submittedCards.filter(
      (card) => card.playerID !== playerId,
    );

    return {
      ...state,
      submittedCards: cardRemoved || state.submittedCards,
    };
  }
  return state;
}

function reducer(state, action) {
  const { type, payload } = action;
  switch (type) {
    case 'JOIN_LOBBY':
      return joinLobby(state);

    case 'UPDATE':
      return update(state, payload);

    case 'ERROR_DISCONNECT':
      return errorDisconnect(state);

    case 'SUBMIT_CARDS':
      return submitCards(state);

    case 'RECEIVE_WHITE_CARDS':
      return receiveWhiteCards(state, payload);

    case 'LOBBY_CLOSED':
      return lobbyClosed(state);

    case 'NO_LOBBY':
      return noLobby(state);

    case 'REMOVE_DISCONNECTED_PLAYERS_CARD':
      return removeDisconnectedPlayersCard(state, payload);

    default:
      return { ...state };
  }
}

export default reducer;
