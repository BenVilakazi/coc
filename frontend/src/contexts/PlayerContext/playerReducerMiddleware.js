import socketInstance from '../../socket/socket';

function joinLobby({ lobbyId, playerName }) {
  socketInstance.joinLobby(lobbyId, playerName);
}

function submitCards(payload) {
  socketInstance.sendMessage({ event: 'player-submit', payload });
}

function previewWinner(payload) {
  socketInstance.sendMessage({ event: 'preview-winner', payload });
}

function submitWinner(payload) {
  socketInstance.sendMessage({ event: 'winner-selected', payload });
}

export default async function playerReducerMiddleware(
  { type, payload },
  dispatch,
) {
  switch (type) {
    case `JOIN_LOBBY`:
      joinLobby(payload);
      break;

    case `SUBMIT_CARDS`:
      submitCards(payload);
      break;

    case 'PREVIEW_WINNER':
      previewWinner(payload);
      break;

    case 'SUBMIT_WINNER':
      submitWinner(payload);
      break;

    default:
      break;
  }
  dispatch({ type, payload });
}
