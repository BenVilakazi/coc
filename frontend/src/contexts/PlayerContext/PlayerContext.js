import React, { createContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import reducer from './playerReducer';
import socketInstance from '../../socket/socket';
import useReducerMiddleware from '../useReducerMiddleware';
import playerReducerMiddleware from './playerReducerMiddleware';

// when the socket receives messaged from the server, it will fire off an event from this EventEmitter
// importing it allows us to "subscribe" to those events and update our state as needed
const { emitter } = socketInstance;

const initialState = {
  gameState: 'enter-code',
  cards: [],
  message: {
    big: '',
    small: '',
  },
  selectCardCount: 0,
  loading: [],
};

const propTypes = {
  children: PropTypes.node.isRequired,
};

// subscribe to the player context
export const PlayerContext = createContext();

export function PlayerProvider({ children }) {
  const [state, dispatch] = useReducerMiddleware(
    playerReducerMiddleware,
    reducer,
    initialState,
  );
  // this function allows us to parse any incoming messages from the event emitter
  // to make sure we know how to handle them.
  function handleMessage({ event, payload }) {
    switch (event) {
      case 'join-lobby':
        return dispatch({ type: 'JOIN_LOBBY', payload });

      case 'update':
        return dispatch({ type: 'UPDATE', payload });

      case 'error-disconnect':
        return dispatch({ type: 'ERROR_DISCONNECT', payload });

      case 'deal-white-cards':
        return dispatch({ type: 'RECEIVE_WHITE_CARDS', payload });

      case 'lobby-closed':
        return dispatch({ type: 'LOBBY_CLOSED' });

      case 'no-lobby':
        return dispatch({ type: 'NO_LOBBY', payload });

      case 'remove-disconnected-players-card':
        return dispatch({ type: 'REMOVE_DISCONNECTED_PLAYERS_CARD', payload });

      default:
        return undefined;
    }
  }

  useEffect(() => {
    // subscribe to the event emitter
    emitter.on('message', handleMessage);
    // cleanup the listener when dismounting
    return () => {
      emitter.off('message', handleMessage);
    };
  }, []);

  return (
    <PlayerContext.Provider value={{ state, dispatch }}>
      {children}
    </PlayerContext.Provider>
  );
}

PlayerProvider.propTypes = propTypes;
