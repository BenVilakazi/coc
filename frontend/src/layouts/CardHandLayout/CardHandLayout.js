import React, { useContext } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { PlayerContext } from '../../contexts/PlayerContext/PlayerContext';

import Button from '../../components/Buttons/Button';
import config from '../../config';

const propTypes = {
  title: PropTypes.shape({
    top: PropTypes.string.isRequired,
    bottom: PropTypes.string.isRequired,
  }).isRequired,
  children: PropTypes.node.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onClear: PropTypes.func.isRequired,
  numberSelected: PropTypes.number.isRequired,
};

const { largeMobileWidth } = config.breakpoint.playerBreakpoints;

const LayoutContainer = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;

  display: flex;
  flex-direction: column;

  .card-container {
    display: flex;
    max-width: 100%;
    margin: auto 0 auto 0;
    overflow: auto;
    padding-left: 20px;
  }

  .card-container > * {
    margin-right: 20px;
    flex-shrink: 0;
  }

  @media (min-height: 640px) and (min-width: ${largeMobileWidth}) {
    .card-container {
      flex-wrap: wrap;
      justify-content: center;
      padding-top: 40px;
    }

    .card-container > * {
      margin-bottom: 40px;
    }
  }
`;

const Header = styled.header`
  width: 100%;
  min-height: 85px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);

  display: flex;
  justify-content: space-between;

  .header-txt {
    margin: 10px 25px -6px -2px;
    line-height: 200%;
    align-self: flex-end;
    font-size: 18px;
  }

  h1 span {
    display: block;
  }

  .button-container {
    display: flex;
    flex-direction: row;
    align-items: center;
  }

  .button {
    text-align: center;
    padding: 0 10px 0 10px;
    height: 30px;
    width: 90px;
    margin-right: 15px;
    border: none;
  }

  @media (min-height: 640px) and (min-width: ${largeMobileWidth}) {
    min-height: 100px;

    .header-txt {
      margin: 26px 70px -7px -2px;
      font-size: 1.3rem;
    }

    .button-container {
      margin-right: 40px;
    }
  }

  @media (max-width: 396px) {
    .header-txt {
      margin: 10px;
      margin-left: 0;
      font-size: 12px;
      margin-bottom: -4px;
    }
  }

  @media (min-width: 397px) and (max-width: 500px) {
    .header-txt {
      margin: 10px;
      margin-left: 0;
      font-size: 18px;
      margin-bottom: -7px;
    }
  }
`;

export default function CardHandLayout({
  title,
  children,
  onSubmit,
  onClear,
  numberSelected,
}) {
  const { state } = useContext(PlayerContext);

  const areAllCardsSelected = numberSelected === state.selectCardCount;

  function submitButtonText() {
    return areAllCardsSelected
      ? 'SUBMIT'
      : `${numberSelected}/${state.selectCardCount} SELECTED`;
  }

  return (
    <LayoutContainer>
      <Header className="header-container">
        <div className="header-txt">
          <h1>
            {title.top}
            <span>{title.bottom}</span>
          </h1>
        </div>

        <section className="button-container">
          <Button
            isActive={areAllCardsSelected}
            disabled={!areAllCardsSelected}
            onClick={onSubmit}
            className="button"
            data-testid="submit"
          >
            {submitButtonText()}
          </Button>
          <Button onClick={onClear} className="button" isActive>
            CLEAR
          </Button>
        </section>
      </Header>

      <div className="card-container">{children}</div>
    </LayoutContainer>
  );
}

CardHandLayout.propTypes = propTypes;
