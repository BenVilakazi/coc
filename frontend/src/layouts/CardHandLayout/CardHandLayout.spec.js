import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import CardHandLayout from './CardHandLayout';
import { PlayerContext } from '../../contexts/PlayerContext/PlayerContext';

describe('CardHandLayout', () => {
  describe('rendering', () => {
    it('renders the submit button', () => {
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(
        screen.getByRole('button', { name: 'SUBMIT' }),
      ).toBeInTheDocument();
    });

    it('renders the clear button', () => {
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(screen.getByRole('button', { name: 'CLEAR' })).toBeInTheDocument();
    });

    it('renders the title', () => {
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(screen.getByText('test')).toBeInTheDocument();
      expect(screen.getByText('bar')).toBeInTheDocument();
    });

    it('renders children', () => {
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            <p>child1</p>
            <p>child2</p>
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(screen.getByText('child1')).toBeInTheDocument();
      expect(screen.getByText('child2')).toBeInTheDocument();
    });

    it('matches the snapshot', () => {
      const state = { selectCardCount: 1 };

      const tree = renderer
        .create(
          <PlayerContext.Provider value={{ state }}>
            <CardHandLayout
              title={{ top: 'test', bottom: 'bar' }}
              onSubmit={() => {}}
              onClear={() => {}}
              numberSelected={1}
            >
              FOO
            </CardHandLayout>
            ,
          </PlayerContext.Provider>,
        )
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('functionality', () => {
    it('does not call the onSubmit callback when the SUBMIT button is not clicked', () => {
      const mockSubmit = jest.fn();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={mockSubmit}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(mockSubmit).not.toHaveBeenCalledTimes(1);
    });

    it('does not call the onSubmit callback if the submit button is pressed and there are not enough cards selected', () => {
      const mockSubmit = jest.fn();
      const state = { selectCardCount: 2 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={mockSubmit}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      userEvent.click(screen.getByTestId('submit'));
      expect(mockSubmit).not.toHaveBeenCalled();
    });

    it('calls the onSubmit callback when the SUBMIT button is clicked', () => {
      const mockSubmit = jest.fn();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={mockSubmit}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      userEvent.click(screen.getByRole('button', { name: 'SUBMIT' }));
      expect(mockSubmit).toHaveBeenCalledTimes(1);
    });

    it('does not call the onClear callback when the CLEAR button is not clicked', () => {
      const mockClear = jest.fn();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
            onClear={mockClear}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(mockClear).not.toHaveBeenCalledTimes(1);
    });

    it('calls the onClear callback when the CLEAR button is clicked', () => {
      const mockClear = jest.fn();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
            onClear={mockClear}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      userEvent.click(screen.getByRole('button', { name: 'CLEAR' }));
      expect(mockClear).toHaveBeenCalledTimes(1);
    });
  });

  describe('propTypes', () => {
    it('logs an error when onClear is not included', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onSubmit={() => {}}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when onClear is not a function', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onClear="Not a function"
            onSubmit={() => {}}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when onSubmit is not included', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when onSubmit is not a function', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'test', bottom: 'bar' }}
            onClear={() => {}}
            numberSelected={1}
            onSubmit="Not a function"
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error when title is not included', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      expect(() => {
        render(
          <PlayerContext.Provider value={{ state }}>
            <CardHandLayout
              onClear={() => {}}
              numberSelected={1}
              onSubmit={() => {}}
            >
              FOO
            </CardHandLayout>
            ,
          </PlayerContext.Provider>,
        );
      }).toThrow();

      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error if title.top is not included', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ bottom: 'string' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error if title.bottom is not included', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'string' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error if title.top is not passed a string', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 10, bottom: 'string' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error if title.bottom is not passed a string', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'string', bottom: 10 }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          >
            FOO
          </CardHandLayout>
          ,
        </PlayerContext.Provider>,
      );
      expect(consoleSpy).toHaveBeenCalled();
    });

    it('logs an error if children are not passed in', () => {
      const consoleSpy = jest.spyOn(console, 'error').mockImplementation();
      const state = { selectCardCount: 1 };

      render(
        <PlayerContext.Provider value={{ state }}>
          <CardHandLayout
            title={{ top: 'line1', bottom: 'string' }}
            onSubmit={() => {}}
            onClear={() => {}}
            numberSelected={1}
          />
          ,
        </PlayerContext.Provider>,
      );
      expect(consoleSpy).toHaveBeenCalled();
    });
  });
});
