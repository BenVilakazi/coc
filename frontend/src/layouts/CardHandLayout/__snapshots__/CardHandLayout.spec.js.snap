// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardHandLayout rendering matches the snapshot 1`] = `
Array [
  .c2 {
  display: block;
  border: 2px solid var(--primary-text-color);
  font-size: 12px;
  font-weight: bold;
  padding: 0.7em;
  background: var(--secondary-background-color);
  color: var(--secondary-text-color);
  background: var(--primary-background-color);
  color: var(--primary-text-color);
}

.c2:hover,
.c2:active,
.c2:focus {
  background: var(--accent-background-color);
  cursor: pointer;
}

.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .card-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-width: 100%;
  margin: auto 0 auto 0;
  overflow: auto;
  padding-left: 20px;
}

.c0 .card-container > * {
  margin-right: 20px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c1 {
  width: 100%;
  min-height: 85px;
  background-color: var(--secondary-background-color);
  color: var(--secondary-text-color);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 .header-txt {
  margin: 10px 25px -6px -2px;
  line-height: 200%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  font-size: 18px;
}

.c1 h1 span {
  display: block;
}

.c1 .button-container {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .button {
  text-align: center;
  padding: 0 10px 0 10px;
  height: 30px;
  width: 90px;
  margin-right: 15px;
  border: none;
}

@media (min-height:640px) and (min-width:600px) {
  .c0 .card-container {
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    padding-top: 40px;
  }

  .c0 .card-container > * {
    margin-bottom: 40px;
  }
}

@media (min-height:640px) and (min-width:600px) {
  .c1 {
    min-height: 100px;
  }

  .c1 .header-txt {
    margin: 26px 70px -7px -2px;
    font-size: 1.3rem;
  }

  .c1 .button-container {
    margin-right: 40px;
  }
}

@media (max-width:396px) {
  .c1 .header-txt {
    margin: 10px;
    margin-left: 0;
    font-size: 12px;
    margin-bottom: -4px;
  }
}

@media (min-width:397px) and (max-width:500px) {
  .c1 .header-txt {
    margin: 10px;
    margin-left: 0;
    font-size: 18px;
    margin-bottom: -7px;
  }
}

<div
    className="c0"
  >
    <header
      className="c1 header-container"
    >
      <div
        className="header-txt"
      >
        <h1>
          test
          <span>
            bar
          </span>
        </h1>
      </div>
      <section
        className="button-container"
      >
        <button
          className="c2 button"
          data-testid="submit"
          disabled={false}
          onClick={[Function]}
        >
          SUBMIT
        </button>
        <button
          className="c2 button"
          onClick={[Function]}
        >
          CLEAR
        </button>
      </section>
    </header>
    <div
      className="card-container"
    >
      FOO
    </div>
  </div>,
  ",",
]
`;
